hazelcast/src/main/java/com/hazelcast/util/FilteringClassLoader.java

```diff
@@ -103,30 +103,31 @@ public InputStream getResourceAsStream(String name) {
                     return clazz;
                 }
             } finally {
-                IOUtil.closeResource(classLoadingMutex);
+                closeResource(classLoadingMutex);
             }
         }
         return delegatingClassLoader.loadClass(name);
     }
 
     private Class<?> loadAndDefineClass(String name) throws ClassNotFoundException {
-        InputStream is = getResourceAsStream(name.replaceAll("\\.", "/").concat(".class"));
+        InputStream is = null;
+        ByteArrayOutputStream os = null;
         try {
-            byte[] temp = new byte[BUFFER_SIZE];
-            ByteArrayOutputStream baos = new ByteArrayOutputStream();
+            is = getResourceAsStream(pattern.matcher(name).replaceAll("/").concat(".class"));
+            os = new ByteArrayOutputStream();
 
             int length;
-            while ((length = is.read(temp)) != -1) {
-                baos.write(temp, 0, length);
+            while ((length = is.read(buffer)) != -1) {
+                os.write(buffer, 0, length);
             }
 
-            byte[] data = baos.toByteArray();
-            Class<?> clazz = defineClass(name, data, 0, data.length);
-            return clazz;
+            byte[] data = os.toByteArray();
+            return defineClass(name, data, 0, data.length);
         } catch (Exception e) {
             throw new ClassNotFoundException(name, e);
         } finally {
-            IOUtil.closeResource(is);
+            closeResource(os);
+            closeResource(is);
```

Comment:
In theory the output stream doesn't need to be closed since ByteArrayOutputStream doesn't hold any resources unlike e.g. a FileOutputStream. But it is clean anway.